File: sdks/go/path_utils.go
package mdp

import (
	"regexp"
	"strings"
)

var staticSegmentPattern = regexp.MustCompile(`^(?:[a-z0-9](?:[a-z0-9_-]*)?|\.[a-z0-9](?:[a-z0-9_-]*)?)$`)
var paramSegmentPattern = regexp.MustCompile(`^:[a-z0-9](?:[a-z0-9_-]*)$`)

var reservedLeafNames = map[string]struct{}{
	"skill.md":  {},
	"prompt.md": {},
	"SKILL.md":  {},
	"PROMPT.md": {},
}

type PathPatternMatch struct {
	Params      map[string]any
	Specificity []int
}

func IsPathPattern(value string) bool {
	return validatePath(value, true)
}

func IsConcretePath(value string) bool {
	return validatePath(value, false)
}

func IsSkillPath(value string) bool {
	if !IsPathPattern(value) {
		return false
	}
	segments := splitPath(value)
	if len(segments) == 0 {
		return false
	}
	leaf := segments[len(segments)-1]
	return leaf == "skill.md" || leaf == "SKILL.md"
}

func IsPromptPath(value string) bool {
	if !IsPathPattern(value) {
		return false
	}
	segments := splitPath(value)
	if len(segments) == 0 {
		return false
	}
	leaf := segments[len(segments)-1]
	return leaf == "prompt.md" || leaf == "PROMPT.md"
}

func MatchPathPattern(pattern string, path string) *PathPatternMatch {
	if !IsPathPattern(pattern) || !IsConcretePath(path) {
		return nil
	}

	patternSegments := splitPath(pattern)
	pathSegments := splitPath(path)
	if len(patternSegments) != len(pathSegments) {
		return nil
	}

	params := map[string]any{}
	specificity := make([]int, 0, len(patternSegments))
	for index := range patternSegments {
		patternSegment := patternSegments[index]
		pathSegment := pathSegments[index]

		if strings.HasPrefix(patternSegment, ":") {
			params[strings.TrimPrefix(patternSegment, ":")] = pathSegment
			specificity = append(specificity, 0)
			continue
		}

		if patternSegment != pathSegment {
			return nil
		}

		if isReservedLeafName(patternSegment) {
			specificity = append(specificity, 1)
		} else {
			specificity = append(specificity, 2)
		}
	}

	return &PathPatternMatch{
		Params:      params,
		Specificity: specificity,
	}
}

func ComparePathSpecificity(left []int, right []int) int {
	maxLength := len(left)
	if len(right) > maxLength {
		maxLength = len(right)
	}

	for index := 0; index < maxLength; index += 1 {
		leftValue := -1
		rightValue := -1
		if index < len(left) {
			leftValue = left[index]
		}
		if index < len(right) {
			rightValue = right[index]
		}
		if leftValue != rightValue {
			return leftValue - rightValue
		}
	}

	return 0
}

func validatePath(value string, allowParams bool) bool {
	if !strings.HasPrefix(value, "/") || strings.Contains(value, "?") || strings.Contains(value, "#") {
		return false
	}

	segments := splitPath(value)
	if len(segments) == 0 {
		return false
	}

	for index, segment := range segments {
		if segment == "" {
			return false
		}
		isLast := index == len(segments)-1
		if isReservedLeafName(segment) {
			if !isLast {
				return false
			}
			continue
		}
		if allowParams && paramSegmentPattern.MatchString(segment) {
			continue
		}
		if !staticSegmentPattern.MatchString(segment) {
			return false
		}
	}

	return true
}

func splitPath(value string) []string {
	return strings.Split(strings.TrimPrefix(value, "/"), "/")
}

func isReservedLeafName(value string) bool {
	_, ok := reservedLeafNames[value]
	return ok
}

File: sdks/rust/src/path_utils.rs
use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::{Map, Value};

static STATIC_SEGMENT_PATTERN: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"^(?:[a-z0-9](?:[a-z0-9_-]*)?|\.[a-z0-9](?:[a-z0-9_-]*)?)$").unwrap()
});
static PARAM_SEGMENT_PATTERN: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"^:[a-z0-9](?:[a-z0-9_-]*)$").unwrap());
static RESERVED_LEAF_NAMES: &[&str] = &["skill.md", "prompt.md", "SKILL.md", "PROMPT.md"];

#[derive(Clone, Debug, PartialEq)]
pub struct PathPatternMatch {
    pub params: Map<String, Value>,
    pub specificity: Vec<i32>,
}

pub fn is_path_pattern(value: &str) -> bool {
    validate_path(value, true)
}

pub fn is_concrete_path(value: &str) -> bool {
    validate_path(value, false)
}

pub fn is_skill_path(value: &str) -> bool {
    is_path_pattern(value)
        && split_path(value)
            .last()
            .map(|segment| *segment == "skill.md" || *segment == "SKILL.md")
            .unwrap_or(false)
}

pub fn is_prompt_path(value: &str) -> bool {
    is_path_pattern(value)
        && split_path(value)
            .last()
            .map(|segment| *segment == "prompt.md" || *segment == "PROMPT.md")
            .unwrap_or(false)
}

pub fn match_path_pattern(pattern: &str, path: &str) -> Option<PathPatternMatch> {
    if !is_path_pattern(pattern) || !is_concrete_path(path) {
        return None;
    }

    let pattern_segments = split_path(pattern);
    let path_segments = split_path(path);
    if pattern_segments.len() != path_segments.len() {
        return None;
    }

    let mut params = Map::new();
    let mut specificity = Vec::with_capacity(pattern_segments.len());

    for (pattern_segment, path_segment) in pattern_segments.iter().zip(path_segments.iter()) {
        if let Some(param_name) = pattern_segment.strip_prefix(':') {
            params.insert(param_name.to_string(), Value::String((*path_segment).to_string()));
            specificity.push(0);
            continue;
        }

        if pattern_segment != path_segment {
            return None;
        }

        specificity.push(if is_reserved_leaf_name(pattern_segment) { 1 } else { 2 });
    }

    Some(PathPatternMatch { params, specificity })
}

pub fn compare_path_specificity(left: &[i32], right: &[i32]) -> i32 {
    let max_length = left.len().max(right.len());
    for index in 0..max_length {
        let left_value = left.get(index).copied().unwrap_or(-1);
        let right_value = right.get(index).copied().unwrap_or(-1);
        if left_value != right_value {
            return left_value - right_value;
        }
    }
    0
}

fn validate_path(value: &str, allow_params: bool) -> bool {
    if !value.starts_with('/') || value.contains('?') || value.contains('#') {
        return false;
    }

    let segments = split_path(value);
    if segments.is_empty() {
        return false;
    }

    for (index, segment) in segments.iter().enumerate() {
        if segment.is_empty() {
            return false;
        }

        let is_last = index == segments.len() - 1;
        if is_reserved_leaf_name(segment) {
            if !is_last {
                return false;
            }
            continue;
        }

        if allow_params && PARAM_SEGMENT_PATTERN.is_match(segment) {
            continue;
        }

        if !STATIC_SEGMENT_PATTERN.is_match(segment) {
            return false;
        }
    }

    true
}

fn split_path(value: &str) -> Vec<&str> {
    value.split('/').skip(1).collect()
}

fn is_reserved_leaf_name(value: &str) -> bool {
    RESERVED_LEAF_NAMES.contains(&value)
}
